logsrvd/logsrvd.c

```diff
@@ -784,6 +825,309 @@ signal_cb(int signo, int what, void *v)
+    }
+
+    x509 = PEM_read_bio_X509_AUX(cert, NULL, NULL, NULL);
+
+exit:
+    if (cert)
+        BIO_free(cert);
+
+    debug_return_ptr(x509);
+}
+
+static bool
+check_cert(X509_STORE *ca_store_ctx, SSL_CTX *ctx, const char *cert_file)
+{
+    bool ret = false;
+    X509_STORE_CTX *store_ctx = NULL;
+    X509 *x509 = NULL;
+
+    debug_decl(check_cert, SUDO_DEBUG_UTIL)
+
+    if ((x509 = load_cert(cert_file)) == NULL)
+        goto exit;
+
+    if ((store_ctx = X509_STORE_CTX_new()) == NULL) {
+        sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+            "unable to allocate X509_STORE_CTX object: %s",
+            ERR_error_string(ERR_get_error(), NULL));
+        goto exit;
+    }
+
+    X509_STORE_set_flags(ca_store_ctx, X509_V_FLAG_X509_STRICT);
+
+    if (!X509_STORE_CTX_init(store_ctx, ca_store_ctx, x509, 0)) {
+        sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+            "unable to initialize X509_STORE_CTX object: %s",
+            ERR_error_string(ERR_get_error(), NULL));
+        goto exit;
+    }
+
+    if (X509_verify_cert(store_ctx) <= 0) {
+        sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+            "unable to verify cert: %s",
+            ERR_error_string(ERR_get_error(), NULL));
+        goto exit;
+    }
+
+    /* everything is good, use this server certificate during TLS handshakes */
+    if (!SSL_CTX_use_certificate(ctx, x509)) {
+        sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+            "unable to load cert to the ssl context: %s",
+            ERR_error_string(ERR_get_error(), NULL));
+        goto exit;
+    }
+
+
+    ret = true;
+exit:
+    if (store_ctx)
+        X509_STORE_CTX_free(store_ctx);
+
+    if (x509)
+        X509_free(x509);
+
+    debug_return_bool(ret);
+}
+
+static bool
+verify_server_cert(SSL_CTX *ctx, const struct logsrvd_tls_config *tls_config)
+{
+    bool ret = false;
+    X509_STORE *ca_store_ctx = NULL;
+    X509_LOOKUP *x509_lookup = NULL;
+
+    debug_decl(verify_server_cert, SUDO_DEBUG_UTIL)
+
+    if ((ca_store_ctx = X509_STORE_new()) == NULL) {
+        sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+            "unable to allocate X509_STORE object for CA bundle: %s",
+            ERR_error_string(ERR_get_error(), NULL));
+        goto exit;
+    }
+
+    if ((x509_lookup = X509_STORE_add_lookup(ca_store_ctx, X509_LOOKUP_file())) == NULL) {
+        sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+            "unable to set lookup method: %s",
+            ERR_error_string(ERR_get_error(), NULL));
+        goto exit;
+    }
+
+    if(!X509_LOOKUP_load_file(x509_lookup, tls_config->cacert_path, X509_FILETYPE_PEM)) {
+        sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+            "unable to load CA bundle file: %s",
+            ERR_error_string(ERR_get_error(), NULL));
+        goto exit;
+    }
+
+    ret = check_cert(ca_store_ctx, ctx, tls_config->cert_path);
+
+exit:
+    if (!ca_store_ctx)
```

Comment:
same here,  X509_STORE_free() already checks for a NULL pointer.

Reply:
fixed